File: backend/src/main/java/br/com/calciolari/datahub/imports/infrastructure/interpdv/qrp/EmfTextRecordExtractor.java
package br.com.calciolari.datahub.imports.infrastructure.interpdv.qrp;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Walks EMF records inside a page and extracts {@code EMR_EXTTEXTOUTW} text.
 *
 * <p>Evidence from PoC {@code parseEmfTexts}:
 * <ul>
 *   <li>record header: type u32 LE @0, size u32 LE @4</li>
 *   <li>type {@code 84} = EMR_EXTTEXTOUTW; requires {@code size >= 76}</li>
 *   <li>x i32 @36, y i32 @40, chars u32 @44, off u32 @48 (relative to record start)</li>
 *   <li>text UTF-16LE, {@code chars} code units</li>
 *   <li>type {@code 14} = EMR_EOF stops the walk</li>
 * </ul>
 */
public final class EmfTextRecordExtractor {

	static final int EMR_EOF = 14;
	static final int EMR_EXTTEXTOUTW = 84;
	static final int MIN_EXTTEXTOUTW_SIZE = 76;
	static final int MAX_CHARS = 10_000;
	static final int MAX_RECORDS_PER_PAGE = 100_000;

	private final int maxRecordsPerPage;
	private final int maxChars;

	public EmfTextRecordExtractor() {
		this(MAX_RECORDS_PER_PAGE, MAX_CHARS);
	}

	public EmfTextRecordExtractor(int maxRecordsPerPage, int maxChars) {
		if (maxRecordsPerPage < 1) {
			throw new IllegalArgumentException("maxRecordsPerPage must be >= 1");
		}
		if (maxChars < 1) {
			throw new IllegalArgumentException("maxChars must be >= 1");
		}
		this.maxRecordsPerPage = maxRecordsPerPage;
		this.maxChars = maxChars;
	}

	public List<EmfTextRun> extract(byte[] bytes, EmfPage page, int pageIndex) {
		Objects.requireNonNull(bytes, "bytes");
		Objects.requireNonNull(page, "page");
		if (page.endExclusive() > bytes.length) {
			throw new IllegalArgumentException("page exceeds buffer");
		}

		ByteBuffer view = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN);
		List<EmfTextRun> out = new ArrayList<>();
		int pos = page.start();
		int end = page.endExclusive();
		int guard = 0;

		while (pos + 8 <= end && guard++ < maxRecordsPerPage) {
			int type = view.getInt(pos);
			long sizeLong = Integer.toUnsignedLong(view.getInt(pos + 4));
			if (sizeLong < 8 || sizeLong > Integer.MAX_VALUE) {
				break;
			}
			int size = (int) sizeLong;
			if (pos > end - size) {
				break;
			}

			if (type == EMR_EXTTEXTOUTW && size >= MIN_EXTTEXTOUTW_SIZE) {
				int x = view.getInt(pos + 36);
				int y = view.getInt(pos + 40);
				long charsLong = Integer.toUnsignedLong(view.getInt(pos + 44));
				long offLong = Integer.toUnsignedLong(view.getInt(pos + 48));
				if (charsLong <= maxChars && offLong <= Integer.MAX_VALUE && charsLong <= Integer.MAX_VALUE / 2) {
					int chars = (int) charsLong;
					int off = (int) offLong;
					int from = pos + off;
					int to = from + chars * 2;
					if (from >= pos && to >= from && to <= pos + size) {
						String text = decodeUtf16Le(bytes, from, to);
						if (!text.isEmpty()) {
							out.add(new EmfTextRun(pageIndex, x, y, text));
						}
					}
				}
			}

			pos += size;
			if (type == EMR_EOF) {
				break;
			}
		}
		return List.copyOf(out);
	}

	private static String decodeUtf16Le(byte[] bytes, int from, int to) {
		String raw = new String(bytes, from, to - from, StandardCharsets.UTF_16LE);
		return raw.replace("\u0000", "").trim();
	}
}

File: backend/src/Calciolari.DataHub/Imports/Infrastructure/InterPdv/Qrp/EmfTextRecordExtractor.cs
using System.Buffers.Binary;
using System.Text;

namespace Calciolari.DataHub.Imports.Infrastructure.InterPdv.Qrp;

/// <summary>
/// Walks EMF records inside a page and extracts EMR_EXTTEXTOUTW text.
/// Evidence from PoC parseEmfTexts.
/// </summary>
public sealed class EmfTextRecordExtractor
{
    internal const int EmrEof = 14;
    internal const int EmrExtTextOutW = 84;
    internal const int MinExtTextOutWSize = 76;
    internal const int MaxChars = 10_000;
    internal const int MaxRecordsPerPage = 100_000;

    private readonly int _maxRecordsPerPage;
    private readonly int _maxChars;

    public EmfTextRecordExtractor()
        : this(MaxRecordsPerPage, MaxChars)
    {
    }

    public EmfTextRecordExtractor(int maxRecordsPerPage, int maxChars)
    {
        if (maxRecordsPerPage < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(maxRecordsPerPage), "maxRecordsPerPage must be >= 1");
        }

        if (maxChars < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(maxChars), "maxChars must be >= 1");
        }

        _maxRecordsPerPage = maxRecordsPerPage;
        _maxChars = maxChars;
    }

    public IReadOnlyList<EmfTextRun> Extract(byte[] bytes, EmfPage page, int pageIndex)
    {
        ArgumentNullException.ThrowIfNull(bytes);
        ArgumentNullException.ThrowIfNull(page);
        if (page.EndExclusive > bytes.Length)
        {
            throw new ArgumentException("page exceeds buffer", nameof(page));
        }

        var output = new List<EmfTextRun>();
        var pos = page.Start;
        var end = page.EndExclusive;
        var guard = 0;

        while (pos + 8 <= end && guard++ < _maxRecordsPerPage)
        {
            var type = BinaryPrimitives.ReadInt32LittleEndian(bytes.AsSpan(pos, 4));
            var sizeLong = BinaryPrimitives.ReadUInt32LittleEndian(bytes.AsSpan(pos + 4, 4));
            if (sizeLong < 8 || sizeLong > int.MaxValue)
            {
                break;
            }

            var size = (int)sizeLong;
            if (pos > end - size)
            {
                break;
            }

            if (type == EmrExtTextOutW && size >= MinExtTextOutWSize)
            {
                var x = BinaryPrimitives.ReadInt32LittleEndian(bytes.AsSpan(pos + 36, 4));
                var y = BinaryPrimitives.ReadInt32LittleEndian(bytes.AsSpan(pos + 40, 4));
                var charsLong = BinaryPrimitives.ReadUInt32LittleEndian(bytes.AsSpan(pos + 44, 4));
                var offLong = BinaryPrimitives.ReadUInt32LittleEndian(bytes.AsSpan(pos + 48, 4));
                if (charsLong <= (uint)_maxChars && offLong <= int.MaxValue && charsLong <= int.MaxValue / 2)
                {
                    var chars = (int)charsLong;
                    var off = (int)offLong;
                    var from = pos + off;
                    var to = from + chars * 2;
                    if (from >= pos && to >= from && to <= pos + size)
                    {
                        var text = DecodeUtf16Le(bytes, from, to);
                        if (text.Length > 0)
                        {
                            output.Add(new EmfTextRun(pageIndex, x, y, text));
                        }
                    }
                }
            }

            pos += size;
            if (type == EmrEof)
            {
                break;
            }
        }

        return output;
    }

    private static string DecodeUtf16Le(byte[] bytes, int from, int to)
    {
        var raw = Encoding.Unicode.GetString(bytes, from, to - from);
        return raw.Replace("\u0000", string.Empty, StringComparison.Ordinal).Trim();
    }
}
